compiler/resolution/lifetime.cpp

```diff
@@ -0,0 +1,1417 @@
+/*
+ * Copyright 2004-2018 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "lifetime.h"
+
+#include "AstVisitorTraverse.h"
+#include "driver.h"
+#include "expr.h"
+#include "ForLoop.h"
+#include "loopDetails.h"
+#include "stlUtil.h"
+#include "symbol.h"
+#include "view.h"
+#include "wellknown.h"
+
+/* This file implements lifetime checking.
+
+   The lifetime checker infers the lifetime of each symbol (how long what it
+   refers to will be allocated). Globals have infinite lifetime, but a local
+   simple value variable (e.g. an integer) has lifetime of the function block it
+   is declared in.
+
+   The reachability of a symbol is simply the lexical block in which
+   it is alive.
+
+   The checker emits errors for several cases:
+
+     * reachability of a symbol is greater than its lifetime
+     * setting a symbol to another value with longer lifetime
+     * returning a symbol with non-infinite and non-return lifetime
+       (e.g. reference arguments are marked with return lifetime)
+
+   This pass proceeds by computing the minimum lifetime of each
+   local variable and then proceeds to look for the above error cases.
+   Lifetime inference is particularly important to making this
+   system usable.
+
+   * pointer/ref arguments are inferred to be return scope
+     generally, but for methods, the default is that only 'this'
+     is return scope.
+
+   When considering the lifetime of the result of a call,
+   this pass assumes that the lifetime is the minimum of the
+   lifetimes of actual arguments with return scope.
+
+   Records are subject to lifetime checking if they contain
+   fields that are subject to lifetime checking.
+
+   Assumption: = operator overloads for records or classes
+   handle lifetimes as if by-value. But note, ownership
+   operations with Shared or Owned records count as "by value"
+   here.
+
```

Comment:
I suggest adding brief explanations of how this analysis handles aggregates: records, classes, arrays, an array's domain and distribution. Sounds like it happens implicitly by appropriate annotations on certain successor functions?